### PR TITLE
CPLAT-4887: propTypes in Component2

### PR DESCRIPTION
**Pull request by kealjones-wk**
# Depends on https://github.com/cleandart/react-dart/pull/192
# Until @greglittlefield-wf's new adapter changes, so diff will be a little dirty until thats merged.
## Motivation
  
Add propTypes compatibility to Component2 to include Typed `props` arguments.

## Changes
  
- Added `jsifyPropTypes` implementation to the component bridge.
- Added automatic prop checks for props with `@required` annotations
- Updated `validateProps` & `validateRequiredProps` to throw `UnsupportedError` and added `mustCallSuper`
- Added Tests

## Review
_[See CONTRIBUTING.md][contributing-review-types] for more details on review types (+1 / QA +1 / +10) and code review process._

  

Please review: 

### QA Checklist
- [ ] Tests were updated and provide good coverage of the changeset and other affected code
- [ ] Manual testing was performed if needed
    - [ ] Steps from PR author: 
        
    - [ ] Anything falling under manual testing criteria [outlined in CONTRIBUTING.md][contributing-manual-testing]

## Merge Checklist
While we perform many automated checks before auto-merging, some manual checks are needed:
- [ ] A Client Platform member has reviewed these changes
- [ ] There are no unaddressed comments _- this check can be automated if reviewers use the "Request Changes" feature_
- [ ] _For release PRs -_ Version metadata in Rosie comment is correct

[contributing-review-types]: https://github.com/Workiva/over_react/blob/master/CONTRIBUTING.md#review-types
[contributing-manual-testing]: https://github.com/Workiva/over_react/blob/master/CONTRIBUTING.md#manual-testing-criteria

### CODE REVIEW

**Comment by aviary3-wk** [comment]
## Security Insights
No security relevant content was detected by automated scans.

## Action Items
* Review PR for [security impact](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); comment "security review required" if needed or unsure
* Verify [`aviary.yaml`](https://dev.workiva.net/docs/teams/information-security/aviary#configuring-aviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

File: test/over_react/component_declaration/builder_integration_tests/component2/constant_required_accessor_integration_test.dart [hunk 1]
```diff
@@ -11,127 +11,186 @@
...
+        var jacket = mount((ComponentTest()
+            ..required = true
+            ..nullable = true
+          )(),
+          attachedToDocument: true,
+        );
 
-        expect(() => react_dom.render((ComponentTest()..nullable = true)(), mountNode),
-            throwsPropError_Required('ComponentTestProps.required')
+        expect(consoleErrorCalled, isFalse, reason: 'should not have outputted a warning but found: $consoleErrorMessage');
```
Comment (greglittlefield-wf) [inline test/over_react/component_declaration/builder_integration_tests/component2/constant_required_accessor_integration_test.dart]: Just an idea, another way you could do this to simplify your `expect`s, while still giving you high-quality error messages:
```dart
consoleErrors = [];
context['console']['error'] = new JsFunction.withThis((self, message, arg1, arg2, arg3) {
  consoleErrors.add(message);
  originalConsoleError.apply([message], thisArg: self);
});
```

```diff
- expect(consoleErrorCalled, isFalse, reason: 'should not have outputted a warning but found: $consoleErrorMessage');
+expect(consoleErrors, isEmpty, reason: 'should not have outputted a warning');

...

-expect(consoleErrorCalled, isTrue, reason: 'should have outputted a warning');
-expect(consoleErrorMessage, contains('ComponentTestProps.required'));
+expect(consoleErrorMessages, [contains('ComponentTestProps.required')]);
```

File: web/component2/src/demo_components/prop_validation.dart [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+class PropTypesComponent extends UiComponent2<PropTypesProps> {
+  @override
+  render() => Fragment()(Dom.code()('props.twoObjects'), ' currently has ${props.twoObjects.length} objects passed to it.');
+
+  // One variation of what the API could look like.
+  //
+  // This is closer to what it looks like in JS, but might be confusing syntax.
+  @override
+  get propTypes => {
+    propKey((props) => props.twoObjects): (PropTypesProps props, String propName, String componentName, String location, String propFullName) {
```
[the remark this answers is not in the document]
Comment (kealjones-wk) [reply web/component2/src/demo_components/prop_validation.dart]: https://github.com/Workiva/over_react/pull/317#discussion_r301336062 make up your mind.

File: lib/src/component_declaration/component_base/component_base_2.dart [hunk 3]
```diff
@@ -270,6 +273,9 @@ abstract class UiComponent2<TProps extends UiProps> extends react.Component2
   //   END Typed props helpers
   // ----------------------------------------------------------------------
   // ----------------------------------------------------------------------
+
+  @override
+  Map<String, react.PropValidator<TProps>> get propTypes => {};
```
[the remark this answers is not in the document]
Comment (kealjones-wk) [reply lib/src/component_declaration/component_base/component_base_2.dart]: `porps`

File: lib/src/component_declaration/component_base.dart [hunk 4]
```diff
@@ -87,7 +88,50 @@ class UiComponent2BridgeImpl extends Component2BridgeImpl {
 
   @override
   JsMap jsifyPropTypes(covariant UiComponent2 component, Map propTypes) {
-    // todo implement jsifyPropTypes
+    // Add [PropValidator]s for props annotated as required.
+    var newPropTypes = Map.from(propTypes);
+    component.consumedProps?.forEach((ConsumedProps consumedProps) {
+      consumedProps?.props?.forEach((PropDescriptor prop) {
```
[the remark this answers is not in the document]
Comment (kealjones-wk) [reply lib/src/component_declaration/component_base.dart]: there was some error happening here, i dont recall atm but ill try this.

File: lib/src/component_declaration/component_base.dart [hunk 5]
```diff
@@ -87,7 +88,50 @@ class UiComponent2BridgeImpl extends Component2BridgeImpl {
...
+          ) {
+            if (prop.isNullable && props.containsKey(prop.key)) return null;
+            if (!prop.isNullable && props[prop.key] != null) return null;
+
+            if (props[propName] == null) {
+              return new PropError.required(propName, prop.errorMessage);
+            }
+            return null;
+          }
+          newPropTypes.addEntries([MapEntry(prop.key, requiredPropValidator)]);
```
[the remark this answers is not in the document]
Comment (kealjones-wk) [reply lib/src/component_declaration/component_base.dart]: :|

File: lib/src/component_declaration/component_base.dart [hunk 6]
```diff
@@ -87,7 +88,50 @@ class UiComponent2BridgeImpl extends Component2BridgeImpl {
...
+            props,
+            String propName,
+            String componentName,
+            String location,
+            String propFullName,
+          ) {
+            if (prop.isNullable && props.containsKey(prop.key)) return null;
+            if (!prop.isNullable && props[prop.key] != null) return null;
+
+            if (props[propName] == null) {
```
[the remark this answers is not in the document]
Comment (kealjones-wk) [reply lib/src/component_declaration/component_base.dart]: Yeah, just tried this and this if doesn't seem to work the same.

File: lib/src/component_declaration/component_base.dart [hunk 7]
```diff
@@ -87,7 +88,52 @@ class UiComponent2BridgeImpl extends Component2BridgeImpl {
...
   @override
   JsMap jsifyPropTypes(covariant UiComponent2 component, Map propTypes) {
-    // todo implement jsifyPropTypes
+    // Add [PropValidator]s for props annotated as required.
+    var newPropTypes = Map.from(propTypes);
+    component.consumedProps?.forEach((ConsumedProps consumedProps) {
+      consumedProps.props.forEach((PropDescriptor prop) {
+        if (!prop.isRequired) return;
+
+        Error requiredPropValidator(
```
Comment (greglittlefield-wf) [inline lib/src/component_declaration/component_base.dart]: #nit formatting, the rest of this function declaration needs one more indent

File: lib/src/component_declaration/component_base.dart [hunk 8]
```diff
@@ -87,7 +88,52 @@ class UiComponent2BridgeImpl extends Component2BridgeImpl {
...
+          ) {
+            if (prop.isNullable && props.containsKey(prop.key)) return null;
+            if (!prop.isNullable && props[prop.key] != null) return null;
+
+            if (props[propName] == null) {
+              return new PropError.required(propName, prop.errorMessage);
+            }
+            return null;
+          }
+          newPropTypes[prop.key] = requiredPropValidator;
```
Comment (greglittlefield-wf) [inline lib/src/component_declaration/component_base.dart]: Mm, what if there is a consumer prop validator as well as the required one; wouldn't the required one get overridden? Could we add a test for that?

We could either chain these two, or call it within the `requiredPropValidator`.

File: test/over_react/component_declaration/builder_integration_tests/component2/constant_required_accessor_integration_test.dart [hunk 9]
```diff
@@ -11,127 +11,182 @@
...
+        consoleErrors.add(message);
+        originalConsoleError.apply([message], thisArg: self);
+      });
+    });
+
+    tearDown(() {
+      context['console']['error'] = originalConsoleError;
+    });
     group('throwing when a prop is required and not set', () {
+
```
Comment (greglittlefield-wf) [inline test/over_react/component_declaration/builder_integration_tests/component2/constant_required_accessor_integration_test.dart]: #nit extra newline

File: test/over_react/component_declaration/builder_integration_tests/component2/constant_required_accessor_integration_test.dart [hunk 10]
```diff
@@ -11,127 +11,182 @@
...
+      originalConsoleError = context['console']['error'];
+      context['console']['error'] = new JsFunction.withThis((self, message) {
+        consoleErrors.add(message);
+        originalConsoleError.apply([message], thisArg: self);
+      });
+    });
+
+    tearDown(() {
+      context['console']['error'] = originalConsoleError;
+    });
```
Comment (greglittlefield-wf) [inline test/over_react/component_declaration/builder_integration_tests/component2/constant_required_accessor_integration_test.dart]: #nit missing newline after

File: test/test_util/proptypes_tester.dart [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+library test_util.prop_utils;
+
+import 'package:over_react_test/over_react_test.dart';
+import 'package:test/test.dart';
+
+void testPropTypes(dynamic instanceBuilder()) {
```
Comment (greglittlefield-wf) [inline test/test_util/proptypes_tester.dart]: Was this meant to be committed? It's unused and seems unfinished

File: lib/src/component_declaration/component_base.dart (hunk 8 above)
Comment (kealjones-wk) [reply lib/src/component_declaration/component_base.dart]: oooooo good thought! golly

File: test/over_react/component_declaration/builder_integration_tests/component2/constant_required_accessor_integration_test.dart (hunk 10 above)
Comment (kealjones-wk) [reply test/over_react/component_declaration/builder_integration_tests/component2/constant_required_accessor_integration_test.dart]: i put the missing newline over here: https://github.com/Workiva/over_react/pull/317#discussion_r310211798

File: test/test_util/proptypes_tester.dart (hunk 11 above)
Comment (kealjones-wk) [reply test/test_util/proptypes_tester.dart]: yeah, ill remove this, it was my starting to think about how to make a testing utility but we have a seperate ticket for that now.

**Comment by greglittlefield-wf** [comment]
+10, besides dependency override that should be removed

**Comment by kealjones-wk** [comment]
+1 to gregs commits :|

**Comment by kealjones-wk** [comment]
@Workiva/release-management-p

**Comment by kealjones-wk** [comment]
how bout now?
@Workiva/release-management-p

**Pull request merged**